Transport task fails for an .hdf file

1.0.0b1 version, python 2.7

When I submit a particular calculation, an `inp.xml` and `cdn_last.hdf` files need to be copied from local to remote machine via ssh. However, only the `inp.xml` is successfully transported to the remote machine while `cdf_last.hdf` is not. See the output:

```
04/08/2019 03:16:40 PM <22387> aiida.orm.nodes.process.calculation.calcjob.CalcJobNode: [INFO] out folder path 13557f20-a39b-487c-bd1d-f315152c825c
04/08/2019 03:16:40 PM <22387> aiida.orm.nodes.process.calculation.calcjob.CalcJobNode: [INFO] local copy file list [('2038166c-1d15-4354-b859-3673e7551271', u'inp.xml', u'inp.xml'), ('13557f20-a39b-487c-bd1d-f315152c825c', 'cdn_last.hdf', 'cdn.hdf')]
04/08/2019 03:17:11 PM <22387> aiida.engine.processes.calcjobs.tasks: [INFO] uploading calculation<24156>

###########
inp.xml
###########
cdn_last.hdf
04/08/2019 03:17:14 PM <22387> aiida.engine.transports: [ERROR] Exception whilst using transport:
Traceback (most recent call last):
  File "/Users/tsep/Documents/aiida/aiida_core/aiida/engine/transports.py", line 106, in request_transport
    yield transport_request.future
  File "/Users/tsep/Documents/aiida/aiida_core/aiida/engine/processes/calcjobs/tasks.py", line 78, in do_upload
    raise Return(execmanager.upload_calculation(node, transport, calc_info, script_filename))
  File "/Users/tsep/Documents/aiida/aiida_core/aiida/engine/daemon/execmanager.py", line 165, in upload_calculation
    handle.write(data_node.get_object_content(filename))
  File "/Users/tsep/Documents/aiida/aiida_core/aiida/orm/nodes/node.py", line 683, in get_object_content
    return self._repository.get_object_content(key)
  File "/Users/tsep/Documents/aiida/aiida_core/aiida/orm/utils/repository.py", line 137, in get_object_content
    return handle.read()
  File "/Users/tsep/aiidapy/lib/python2.7/codecs.py", line 314, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf8' codec can't decode byte 0x89 in position 0: invalid start byte

```

This part 
```
###########
inp.xml
###########
cdn_last.hdf
```
appeared because I changed `aiida_core/aiida/orm/utils/repository.py` a little bit to track the problem:
```
    def open(self, key, mode='r'):
        """Open a file handle to an object stored under the given key.

        :param key: fully qualified identifier for the object within the repository
        :param mode: the mode under which to open the handle
        """
        print('###########')
        print(key)
        return io.open(self._get_base_folder().get_abs_path(key), mode=mode)
```

Apparently aiida_core tries to open an hdf-file and fails decoding it via utf-8 which did not happen in 1.0.0a4 version. 

Calculation folder on the remote machine contains successfully transported `inp.xml` file:
```
[tseplyaev1@jrl01 0fd5-83ed-461f-92b5-56c05ffb1ca6]$ ls -al
total 20
drwxr-sr-x 3 tseplyaev1 cjias1f 4096 Apr  8 15:17 .
drwxr-sr-x 3 tseplyaev1 cjias1f 4096 Apr  8 15:17 ..
drwxr-sr-x 2 tseplyaev1 cjias1f 4096 Apr  8 15:17 .aiida
-rw-r--r-- 1 tseplyaev1 cjias1f  388 Apr  8 15:17 _aiidasubmit.sh
-rw-r--r-- 1 tseplyaev1 cjias1f 4947 Apr  8 15:17 inp.xml
```

--- a/aiida/backends/tests/orm/data/test_singlefile.py
+++ b/aiida/backends/tests/orm/data/test_singlefile.py
@@ -66,9 +66,9 @@ class TestSinglefileData(AiidaTestCase):
         """Test constructing an instance from filelike instead of filepath."""
         content_original = u'some testing text\nwith a newline'
 
-        with tempfile.NamedTemporaryFile(mode='w+') as handle:
+        with tempfile.NamedTemporaryFile(mode='wb+') as handle:
             basename = os.path.basename(handle.name)
-            handle.write(content_original)
+            handle.write(content_original.encode('utf-8'))
             handle.flush()
             handle.seek(0)
             node = SinglefileData(file=handle)
@@ -91,7 +91,7 @@ class TestSinglefileData(AiidaTestCase):
         """Test constructing an instance from a string."""
         content_original = u'some testing text\nwith a newline'
 
-        with io.StringIO(content_original) as handle:
+        with io.BytesIO(content_original.encode('utf-8')) as handle:
             node = SinglefileData(file=handle)
 
         with node.open() as handle:

--- a/aiida/common/folders.py
+++ b/aiida/common/folders.py
@@ -230,6 +230,9 @@ class Folder(object):  # pylint: disable=useless-object-inheritance
         filename = six.text_type(filename)
         filepath = self.get_abs_path(filename)
 
+        if 'b' in mode:
+            encoding = None
+
         with io.open(filepath, mode=mode, encoding=encoding) as handle:
 
             # In python 2 a string literal can either be of unicode or string (bytes) type. Since we do not know what
@@ -243,6 +246,7 @@ class Folder(object):  # pylint: disable=useless-object-inheritance
                 try:
                     shutil.copyfileobj(utf8reader(filelike), handle)
                 except (UnicodeDecodeError, UnicodeEncodeError):
+                    filelike.seek(0)
                     shutil.copyfileobj(filelike, handle)
             else:
                 shutil.copyfileobj(filelike, handle)

--- a/aiida/engine/daemon/execmanager.py
+++ b/aiida/engine/daemon/execmanager.py
@@ -161,9 +161,11 @@ def upload_calculation(node, transport, calc_info, script_filename, dry_run=Fals
 
         # Note, once #2579 is implemented, use the `node.open` method instead of the named temporary file in
         # combination with the new `Transport.put_object_from_filelike`
-        with NamedTemporaryFile(mode='w+') as handle:
-            handle.write(data_node.get_object_content(filename))
+        # Since the content of the node could potentially be binary, we read the raw bytes and pass them on
+        with NamedTemporaryFile(mode='wb+') as handle:
+            handle.write(data_node.get_object_content(filename, mode='rb'))
             handle.flush()
+            handle.seek(0)
             transport.put(handle.name, target)
 
     if dry_run:

--- a/aiida/orm/nodes/data/singlefile.py
+++ b/aiida/orm/nodes/data/singlefile.py
@@ -97,7 +97,7 @@ class SinglefileData(Data):
             pass
 
         if is_filelike:
-            self.put_object_from_filelike(file, key)
+            self.put_object_from_filelike(file, key, mode='wb')
         else:
             self.put_object_from_file(file, key)
 

--- a/aiida/orm/nodes/node.py
+++ b/aiida/orm/nodes/node.py
@@ -682,12 +682,12 @@ class Node(Entity):
         """
         return self._repository.get_object(key)
 
-    def get_object_content(self, key):
+    def get_object_content(self, key, mode='r'):
         """Return the content of a object identified by key.
 
         :param key: fully qualified identifier for the object within the repository
         """
-        return self._repository.get_object_content(key)
+        return self._repository.get_object_content(key, mode)
 
     def put_object_from_tree(self, path, key=None, contents_only=True, force=False):
         """Store a new object under `key` with the contents of the directory located at `path` on this file system.

--- a/aiida/orm/utils/repository.py
+++ b/aiida/orm/utils/repository.py
@@ -126,12 +126,13 @@ class Repository(object):  # pylint: disable=useless-object-inheritance
 
         return File(filename, FileType.FILE)
 
-    def get_object_content(self, key):
+    def get_object_content(self, key, mode='r'):
         """Return the content of a object identified by key.
 
         :param key: fully qualified identifier for the object within the repository
+        :param mode: the mode under which to open the handle
         """
-        with self.open(key) as handle:
+        with self.open(key, mode=mode) as handle:
             return handle.read()
 
     def put_object_from_tree(self, path, key=None, contents_only=True, force=False):

